# 1 **Import Libraries**

In [1]:
import streamlit as st
import numpy as np
import pandas as pd 
import plotly.express as px


In [2]:
import plotly.io as pio
pio.templates.default = "plotly"
print(pio.templates.default)

plotly


# 2 **Data Loading** 

In [3]:
df = pd.read_csv('retail_store_sales.csv')
df.head()

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied
0,TXN_6867343,CUST_09,Patisserie,Item_10_PAT,18.5,10.0,185.0,Digital Wallet,Online,2024-04-08,True
1,TXN_3731986,CUST_22,Milk Products,Item_17_MILK,29.0,9.0,261.0,Digital Wallet,Online,2023-07-23,True
2,TXN_9303719,CUST_02,Butchers,Item_12_BUT,21.5,2.0,43.0,Credit Card,Online,2022-10-05,False
3,TXN_9458126,CUST_06,Beverages,Item_16_BEV,27.5,9.0,247.5,Credit Card,Online,2022-05-07,NaN
4,TXN_4575373,CUST_05,Food,Item_6_FOOD,12.5,7.0,87.5,Digital Wallet,Online,2022-10-02,False


# 3 **Data Understanding** 

## 3.1 Columns Description



| Column | Description |
|--------|------------|
Transaction ID | A unique identifier for each transaction.
Customer ID | A unique identifier for each customer.
Category | The category of the purchased item.
Item | The name of the purchased item.
Price Per Unit | The static price of a single unit of the item.
Quantity | The quantity of the item purchased.
Total Spent | The total amount spent on the transaction. Calculated as Quantity * Price Per Unit.
Payment Method | The method of payment used.
Location | The location where the transaction occurred.
Transaction Date | The date of the transaction.
Discount Applied | Indicates if a discount was applied to the transaction.



### 🛍️ Product Catalog

#### ⚡ Electric Household Essentials

| 🆔 Item Code | 🔌 Item | 💲 Price |
|--------------|--------|----------|
| Item_1_EHE | 🔌 Blender | 5.0 |
| Item_2_EHE | 🍲 Microwave | 6.5 |
| Item_3_EHE | 🍞 Toaster | 8.0 |
| Item_4_EHE | 🧹 Vacuum Cleaner | 9.5 |
| Item_5_EHE | 🌬️ Air Purifier | 11.0 |
| Item_6_EHE | ☕ Electric Kettle | 12.5 |
| Item_7_EHE | 🍚 Rice Cooker | 14.0 |
| Item_8_EHE | 👕 Iron | 15.5 |
| Item_9_EHE | 🌀 Ceiling Fan | 17.0 |
| Item_10_EHE | 🌪️ Table Fan | 18.5 |
| Item_11_EHE | 💨 Hair Dryer | 20.0 |
| Item_12_EHE | 🔥 Heater | 21.5 |
| Item_13_EHE | 💧 Humidifier | 23.0 |
| Item_14_EHE | 🌫️ Dehumidifier | 24.5 |
| Item_15_EHE | ☕ Coffee Maker | 26.0 |
| Item_16_EHE | ❄️ Portable AC | 27.5 |
| Item_17_EHE | 🍳 Electric Stove | 29.0 |
| Item_18_EHE | 🍲 Pressure Cooker | 30.5 |
| Item_19_EHE | 🍳 Induction Cooktop | 32.0 |
| Item_20_EHE | 🚰 Water Dispenser | 33.5 |
| Item_21_EHE | 🥤 Hand Blender | 35.0 |
| Item_22_EHE | ⚙️ Mixer Grinder | 36.5 |
| Item_23_EHE | 🥪 Sandwich Maker | 38.0 |
| Item_24_EHE | 🍟 Air Fryer | 39.5 |
| Item_25_EHE | 🧃 Juicer | 41.0 |

#### 🪑 Furniture Collection

| 🆔 Item Code | 🛋️ Item | 💲 Price |
|--------------|--------|----------|
| Item_1_FUR | 🪑 Office Chair | 5.0 |
| Item_2_FUR | 🛋️ Sofa | 6.5 |
| Item_3_FUR | ☕ Coffee Table | 8.0 |
| Item_4_FUR | 🍽️ Dining Table | 9.5 |
| Item_5_FUR | 📚 Bookshelf | 11.0 |
| Item_6_FUR | 🛏️ Bed Frame | 12.5 |
| Item_7_FUR | 🚪 Wardrobe | 14.0 |
| Item_8_FUR | 📖 Study Table | 15.5 |
| Item_9_FUR | 🛋️ Recliner | 17.0 |
| Item_10_FUR | 🪑 Side Table | 18.5 |
| Item_11_FUR | 📺 TV Stand | 20.0 |
| Item_12_FUR | 🪞 Dressing Table | 21.5 |
| Item_13_FUR | 🗄️ Chest of Drawers | 23.0 |
| Item_14_FUR | 🧸 Ottoman | 24.5 |
| Item_15_FUR | 🎠 Rocking Chair | 26.0 |
| Item_16_FUR | 👟 Shoe Rack | 27.5 |
| Item_17_FUR | 🌙 Nightstand | 29.0 |
| Item_18_FUR | 🌳 Patio Chair | 30.5 |
| Item_19_FUR | 📐 Folding Table | 32.0 |
| Item_20_FUR | 🛋️ Bean Bag | 33.5 |
| Item_21_FUR | 🪵 Wooden Bench | 35.0 |
| Item_22_FUR | 📦 Corner Shelf | 36.5 |
| Item_23_FUR | 📰 Magazine Rack | 38.0 |
| Item_24_FUR | 🍸 Bar Stool | 39.5 |
| Item_25_FUR | 🪑 Accent Chair | 41.0 |

In [4]:
df.columns


Index(['Transaction ID', 'Customer ID', 'Category', 'Item', 'Price Per Unit',
       'Quantity', 'Total Spent', 'Payment Method', 'Location',
       'Transaction Date', 'Discount Applied'],
      dtype='object')

## 3.2 Feture Types

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              11362 non-null  object 
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Transaction Date  12575 non-null  object 
 10  Discount Applied  8376 non-null   object 
dtypes: float64(3), object(8)
memory usage: 1.1+ MB


In [6]:
# categorical features
cat = ['Customer ID','Category','Item','Payment Method','Location','Discount Applied']

# numerical features
num = ['Price Per Unit','Quantity','Total Spent']

# datetime features
date = ['Transaction Date']

# unwanted features
drop = ['Transaction ID']

## 3.3 Initial Data Inspection

### 3.3.1 General Description

In [7]:
df.describe()

,Price Per Unit,Quantity,Total Spent
count,11966.000000,11971.000000,11971.000000
mean,23.365912,5.536380,129.652577
std,10.743519,2.857883,94.750697
min,5.000000,1.000000,5.000000
25%,14.000000,3.000000,51.000000
50%,23.000000,6.000000,108.500000
75%,33.500000,8.000000,192.000000
max,41.000000,10.000000,410.000000


In [8]:
df.describe(include='object')

,Transaction ID,Customer ID,Category,Item,Payment Method,Location,Transaction Date,Discount Applied
count,12575,12575,12575,11362,12575,12575,12575,8376
unique,12575,25,8,200,3,2,1114,2
top,TXN_6867343,CUST_05,Electric household essentials,Item_2_BEV,Cash,Online,2022-05-30,True
freq,1,544,1591,126,4310,6354,26,4219


### 3.3.2 Missing Values

In [9]:
# total number of missing values
print("Total number of missing values in the dataset:", df.isna().sum().sum())
# percentage of missing values 
print("Percentage of missing values in the dataset:", (df.isna().mean()*100).sum())



Total number of missing values in the dataset: 7229
Percentage of missing values in the dataset: 57.487077534791254


In [10]:
# number of missing values in each column
print("\nNumber of missing values in each column:")
print((df.isna().sum().sort_values(ascending=False)))
# percentage of missing values in each column
print("\nPercentage of missing values in each column:")
print((df.isna().mean()*100).sort_values(ascending=False))


Number of missing values in each column:
Discount Applied    4199
Item                1213
Price Per Unit       609
Total Spent          604
Quantity             604
Transaction ID         0
Customer ID            0
Category               0
Payment Method         0
Location               0
Transaction Date       0
dtype: int64

Percentage of missing values in each column:
Discount Applied    33.391650
Item                 9.646123
Price Per Unit       4.842942
Total Spent          4.803181
Quantity             4.803181
Transaction ID       0.000000
Customer ID          0.000000
Category             0.000000
Payment Method       0.000000
Location             0.000000
Transaction Date     0.000000
dtype: float64


**Findings** <br>
As observed above, the `Total Spent` and `Quantity` columns contain the same number of null values. Consequently, we will check if a null `Quantity` strictly implies a null `Total Spent`, and vice versa.
also, we will check if nulls in `Price Per Unit` correspond to nulls in `Total Spent` or `Quantity`.

**columns with missing values:** [ 'Discount Applied' , 'Item' , 'Price Per Unit' , 'Total Spent' , 'Quantity' ]


In [11]:
df[(df['Total Spent'].isnull()) & (df['Quantity'].isnull())].shape[0]

604

In [12]:
df[((df['Total Spent'].isnull()) | (df['Quantity'].isnull())) & (df['Price Per Unit'].isnull())].shape[0]

0

### 3.3.3 Unique Values

In [13]:
for col in cat:
    print(col)
    print(df[col].nunique())
    print(df[col].unique())
    print('-'*50)

Customer ID
25
['CUST_09' 'CUST_22' 'CUST_02' 'CUST_06' 'CUST_05' 'CUST_07' 'CUST_21'
 'CUST_23' 'CUST_25' 'CUST_14' 'CUST_15' 'CUST_17' 'CUST_01' 'CUST_10'
 'CUST_04' 'CUST_13' 'CUST_18' 'CUST_08' 'CUST_20' 'CUST_12' 'CUST_11'
 'CUST_19' 'CUST_16' 'CUST_24' 'CUST_03']
--------------------------------------------------
Category
8
['Patisserie' 'Milk Products' 'Butchers' 'Beverages' 'Food' 'Furniture'
 'Electric household essentials' 'Computers and electric accessories']
--------------------------------------------------
Item
200
['Item_10_PAT' 'Item_17_MILK' 'Item_12_BUT' 'Item_16_BEV' 'Item_6_FOOD'
 nan 'Item_1_FOOD' 'Item_16_FUR' 'Item_22_BUT' 'Item_3_BUT' 'Item_2_FOOD'
 'Item_24_PAT' 'Item_16_MILK' 'Item_17_PAT' 'Item_13_EHE' 'Item_7_BEV'
 'Item_4_EHE' 'Item_10_FOOD' 'Item_14_FUR' 'Item_20_BUT' 'Item_25_FUR'
 'Item_14_FOOD' 'Item_22_PAT' 'Item_11_FOOD' 'Item_6_PAT' 'Item_21_EHE'
 'Item_25_BEV' 'Item_23_FOOD' 'Item_10_FUR' 'Item_11_BEV' 'Item_23_BUT'
 'Item_22_BEV' 'Item_10_EHE' 'Ite

# 4 **Data Cleaning**

## 4.1 Check for Duplicates

In [14]:
df.duplicated().sum()

np.int64(0)

## 4.2 DateTime Conversion

In [15]:
df['Transaction Date'] = pd.to_datetime(df['Transaction Date'])
#df['Transaction Date'] = df['Transaction Date'].dt.date 
print(df[ 'Transaction Date' ].dtype)

datetime64[ns]


In [16]:
# sort dataset by 'Transaction Date' for better analysis
df = df.sort_values(by='Transaction Date').reset_index(drop=True)
df.head()

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied
0,TXN_2070726,CUST_22,Milk Products,Item_6_MILK,12.5,5.0,62.5,Digital Wallet,In-store,2022-01-01,False
1,TXN_1519499,CUST_15,Computers and electric accessories,Item_6_CEA,12.5,6.0,75.0,Credit Card,Online,2022-01-01,NaN
2,TXN_9364474,CUST_19,Food,Item_22_FOOD,36.5,3.0,109.5,Cash,In-store,2022-01-01,False
3,TXN_6950467,CUST_01,Computers and electric accessories,Item_6_CEA,12.5,6.0,75.0,Digital Wallet,In-store,2022-01-01,True
4,TXN_6329331,CUST_24,Beverages,Item_8_BEV,15.5,5.0,77.5,Credit Card,Online,2022-01-01,True


In [17]:
df['Transaction Date'].describe()

count                            12575
mean     2023-07-12 20:23:41.105368064
min                2022-01-01 00:00:00
25%                2022-09-30 00:00:00
50%                2023-07-13 00:00:00
75%                2024-04-24 00:00:00
max                2025-01-18 00:00:00
Name: Transaction Date, dtype: object

## 4.3 Handle Missing Values

### 4.3.1 Fill `Price Per Unit` using `Total Spent` and `Quantity`

In [18]:
def fill_price_per_unit(row):
    if pd.isna(row['Price Per Unit']):
        if pd.notna(row['Total Spent']) and pd.notna(row['Quantity']):
            return row['Total Spent'] / row['Quantity']
    return row['Price Per Unit']

In [19]:
df['Price Per Unit'] = df.apply(fill_price_per_unit, axis=1)

In [20]:
df.isna().sum().sort_values(ascending=False)

Discount Applied    4199
Item                1213
Total Spent          604
Quantity             604
Category               0
Transaction ID         0
Customer ID            0
Price Per Unit         0
Payment Method         0
Location               0
Transaction Date       0
dtype: int64

### 4.3.2 Fill `Item` using "Product Catalog" from dataset descreiption

In [21]:
category_codes = {
        "Electric household essentials": "EHE",
        "Furniture": "FUR",
        "Patisserie": "PAT",
        "Milk Products": "MILK",
        "Butchers": "BUT",
        "Beverages": "BEV",
        "Food": "FOOD",
        "Computers and electric accessories": "CEA"
    }

In [22]:
def build_price_reference():
    price_reference = []

    base_price = 5.0

    for category, code in category_codes.items():
        for i in range(1, 26):

            item = f'Item_{i}_{code}'
            price = base_price + (i - 1) * 1.5

            price_reference.append([category, item, price])

    price_reference_df = pd.DataFrame(price_reference, columns = ['Category', 'Item Code', 'Price Per Unit'])

    return price_reference_df


In [23]:
price_reference_df = build_price_reference()

In [24]:
price_reference_df.head()

,Category,Item Code,Price Per Unit
0,Electric household essentials,Item_1_EHE,5.0
1,Electric household essentials,Item_2_EHE,6.5
2,Electric household essentials,Item_3_EHE,8.0
3,Electric household essentials,Item_4_EHE,9.5
4,Electric household essentials,Item_5_EHE,11.0


In [25]:
price_reference_df.to_csv('price_reference.csv', index=False)

In [26]:
df = df.merge(price_reference_df, on=['Category', 'Price Per Unit'], how='left')

In [27]:
df.head()

,Transaction ID,Customer ID,Category,Item,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied,Item Code
0,TXN_2070726,CUST_22,Milk Products,Item_6_MILK,12.5,5.0,62.5,Digital Wallet,In-store,2022-01-01,False,Item_6_MILK
1,TXN_1519499,CUST_15,Computers and electric accessories,Item_6_CEA,12.5,6.0,75.0,Credit Card,Online,2022-01-01,NaN,Item_6_CEA
2,TXN_9364474,CUST_19,Food,Item_22_FOOD,36.5,3.0,109.5,Cash,In-store,2022-01-01,False,Item_22_FOOD
3,TXN_6950467,CUST_01,Computers and electric accessories,Item_6_CEA,12.5,6.0,75.0,Digital Wallet,In-store,2022-01-01,True,Item_6_CEA
4,TXN_6329331,CUST_24,Beverages,Item_8_BEV,15.5,5.0,77.5,Credit Card,Online,2022-01-01,True,Item_8_BEV


In [28]:
df.drop(columns = 'Item', inplace = True, errors='ignore')
df.isna().sum().sort_values(ascending=False)

Discount Applied    4199
Quantity             604
Total Spent          604
Category               0
Customer ID            0
Transaction ID         0
Price Per Unit         0
Payment Method         0
Location               0
Transaction Date       0
Item Code              0
dtype: int64

In [29]:
#rename 'Item Code' to 'Item'
df.rename(columns={'Item Code': 'Item'}, inplace=True)
df.head()

,Transaction ID,Customer ID,Category,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied,Item
0,TXN_2070726,CUST_22,Milk Products,12.5,5.0,62.5,Digital Wallet,In-store,2022-01-01,False,Item_6_MILK
1,TXN_1519499,CUST_15,Computers and electric accessories,12.5,6.0,75.0,Credit Card,Online,2022-01-01,NaN,Item_6_CEA
2,TXN_9364474,CUST_19,Food,36.5,3.0,109.5,Cash,In-store,2022-01-01,False,Item_22_FOOD
3,TXN_6950467,CUST_01,Computers and electric accessories,12.5,6.0,75.0,Digital Wallet,In-store,2022-01-01,True,Item_6_CEA
4,TXN_6329331,CUST_24,Beverages,15.5,5.0,77.5,Credit Card,Online,2022-01-01,True,Item_8_BEV


### 4.3.3 Drop missing values in `Total Spent` and `Quantity` columns.

In [30]:
# check percentage of missing values in `Total Spent` and `Quantity` columns
(df[df['Total Spent'].isna() & df['Quantity'].isna()].shape[0]/df.shape[0])*100


4.8031809145129225

Since they are less than 5% of total dataset and cannot be calculated from only `Price Per Unit` column, they will be dropped.

In [31]:
df = df.dropna(subset=['Total Spent', 'Quantity'])
df.isna().sum().sort_values(ascending=False)

Discount Applied    3988
Customer ID            0
Transaction ID         0
Category               0
Price Per Unit         0
Total Spent            0
Quantity               0
Payment Method         0
Location               0
Transaction Date       0
Item                   0
dtype: int64

### 4.3.4 Fill `Discount Applied` with "Unknown"

In [32]:
df['Discount Applied'].fillna('Unknown', inplace=True)

C:\Users\huss1\AppData\Local\Temp\ipykernel_13372\3520738117.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Discount Applied'].fillna('Unknown', inplace=True)


In [33]:
df.head()

,Transaction ID,Customer ID,Category,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied,Item
0,TXN_2070726,CUST_22,Milk Products,12.5,5.0,62.5,Digital Wallet,In-store,2022-01-01,False,Item_6_MILK
1,TXN_1519499,CUST_15,Computers and electric accessories,12.5,6.0,75.0,Credit Card,Online,2022-01-01,Unknown,Item_6_CEA
2,TXN_9364474,CUST_19,Food,36.5,3.0,109.5,Cash,In-store,2022-01-01,False,Item_22_FOOD
3,TXN_6950467,CUST_01,Computers and electric accessories,12.5,6.0,75.0,Digital Wallet,In-store,2022-01-01,True,Item_6_CEA
4,TXN_6329331,CUST_24,Beverages,15.5,5.0,77.5,Credit Card,Online,2022-01-01,True,Item_8_BEV


In [34]:
df.isna().sum()

Transaction ID      0
Customer ID         0
Category            0
Price Per Unit      0
Quantity            0
Total Spent         0
Payment Method      0
Location            0
Transaction Date    0
Discount Applied    0
Item                0
dtype: int64

In [35]:
df.reset_index(drop=True, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11971 entries, 0 to 11970
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    11971 non-null  object        
 1   Customer ID       11971 non-null  object        
 2   Category          11971 non-null  object        
 3   Price Per Unit    11971 non-null  float64       
 4   Quantity          11971 non-null  float64       
 5   Total Spent       11971 non-null  float64       
 6   Payment Method    11971 non-null  object        
 7   Location          11971 non-null  object        
 8   Transaction Date  11971 non-null  datetime64[ns]
 9   Discount Applied  11971 non-null  object        
 10  Item              11971 non-null  object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 1.0+ MB


finally validate data consistency by checking if `Total Spent` is equal to `Price Per Unit` multiplied by `Quantity` for all rows.

In [36]:
(df['Total Spent'] == df['Price Per Unit'] * df['Quantity']).sum()

np.int64(11971)



Up to this point, we have handled missing values using real calculated values rather than statistical imputations (such as mean, median, or mode), and dropped a negligible percentage of rows containing nulls. This approach ensures high data accuracy for the dashboard visualizations.



## 4.4 Drop `Transaction ID` and save cleaned dataset

In [37]:
df = df.drop(columns=['Transaction ID'], errors='ignore')
df.head()

,Customer ID,Category,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied,Item
0,CUST_22,Milk Products,12.5,5.0,62.5,Digital Wallet,In-store,2022-01-01,False,Item_6_MILK
1,CUST_15,Computers and electric accessories,12.5,6.0,75.0,Credit Card,Online,2022-01-01,Unknown,Item_6_CEA
2,CUST_19,Food,36.5,3.0,109.5,Cash,In-store,2022-01-01,False,Item_22_FOOD
3,CUST_01,Computers and electric accessories,12.5,6.0,75.0,Digital Wallet,In-store,2022-01-01,True,Item_6_CEA
4,CUST_24,Beverages,15.5,5.0,77.5,Credit Card,Online,2022-01-01,True,Item_8_BEV


In [38]:
df.to_csv('clean_df.csv', index=False)

In [39]:
clean_df = pd.read_csv('clean_df.csv')
clean_df.head(1)

,Customer ID,Category,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied,Item
0,CUST_22,Milk Products,12.5,5.0,62.5,Digital Wallet,In-store,2022-01-01,False,Item_6_MILK


# 5 **Feature Engineering**

In [40]:
# extract year, month, day of week and month-year from `Transaction Date` for better time-based analysis
clean_df['Transaction Date'] = pd.to_datetime(clean_df['Transaction Date'])
clean_df['year'] = clean_df['Transaction Date'].dt.year
clean_df['month'] = clean_df['Transaction Date'].dt.month
clean_df['day_of_week'] = clean_df['Transaction Date'].dt.day_name()
clean_df['month_year'] = clean_df['Transaction Date'].dt.to_period('M').astype(str)
clean_df.head()

,Customer ID,Category,Price Per Unit,Quantity,Total Spent,Payment Method,Location,Transaction Date,Discount Applied,Item,year,month,day_of_week,month_year
0,CUST_22,Milk Products,12.5,5.0,62.5,Digital Wallet,In-store,2022-01-01,False,Item_6_MILK,2022,1,Saturday,2022-01
1,CUST_15,Computers and electric accessories,12.5,6.0,75.0,Credit Card,Online,2022-01-01,Unknown,Item_6_CEA,2022,1,Saturday,2022-01
2,CUST_19,Food,36.5,3.0,109.5,Cash,In-store,2022-01-01,False,Item_22_FOOD,2022,1,Saturday,2022-01
3,CUST_01,Computers and electric accessories,12.5,6.0,75.0,Digital Wallet,In-store,2022-01-01,True,Item_6_CEA,2022,1,Saturday,2022-01
4,CUST_24,Beverages,15.5,5.0,77.5,Credit Card,Online,2022-01-01,True,Item_8_BEV,2022,1,Saturday,2022-01


# 6 **Exploratory Data Analysis (EDA)**

## 6.1 Initial Exploration

### 6.1.1 Categorical Features

In [41]:
cat

['Customer ID',
 'Category',
 'Item',
 'Payment Method',
 'Location',
 'Discount Applied']

In [42]:
for col in cat:
    fig = px.histogram(clean_df, x=col, title=f'Distribution of {col}')
    fig.update_xaxes(categoryorder='total descending')
    fig.show() 

The distribution of categorical features indicates that sales proportions across all categories are relatively balanced. Furthermore, the purchasing frequency remains consistent across the entire customer base.

### 6.1.2 Numerical Features

In [43]:
for col in num:
    fig = px.histogram(clean_df, x=col, title=f'Distribution of {col}', marginal="box")
    fig.show()

range of unit price --> **5 to 41** --> **uniform** distribution with no outliers <br>
range of quantity --> **1 to 10** --> **uniform** distribution with no outliers <br>
range of total spent --> **5 to 395** --> **right skewed** distribution with small percentage of outliers <br>


### 6.1.3 Time Features

In [44]:
time = ['month_year', 'day_of_week']
for col in time:
    fig = px.histogram(clean_df, x=col, title=f'Distribution of {col}')
    fig.show()

The distribution of time features indicates that sales are consistent across all years, with a slight increase in sales during the months of **January**. Additionally, sales are evenly distributed across all days of the week.

## 6.2 Advanced (High-Value) EDA

In [45]:
# Total Revenue
total_revenue = clean_df['Total Spent'].sum()
print(f'Total Revenue: ${total_revenue:,.2f}')

Total Revenue: $1,552,071.00


### Which category generates the highest revenue?

In [46]:
top_cat = clean_df.groupby(['Discount Applied', 'Category'])['Total Spent'].sum().sort_values(ascending=False).reset_index()
px.histogram(top_cat, x='Category', y='Total Spent', title='Total Spent by Category', color='Discount Applied')


### What are the top-selling items in each category?

In [47]:
# top selling items of each category
top_items = clean_df.groupby(['Category', 'Item'])['Quantity'].sum().sort_values(ascending=False).reset_index()
columns = ['Item', 'Quantity']
top_items_cat = pd.DataFrame(columns=columns)

for category in clean_df['Category'].unique():
    cat_items = top_items[top_items['Category'] == category].sort_values(by='Quantity', ascending=False).head(1)
    top_items_cat = pd.concat([top_items_cat, cat_items], ignore_index=True)
top_items_cat = top_items_cat.sort_values(by='Quantity', ascending=False)
px.bar(top_items_cat, x='Item', y='Quantity', title='Top Selling Item by Category', color='Category')

C:\Users\huss1\AppData\Local\Temp\ipykernel_13372\119363252.py:8: FutureWarning:

The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.



In [48]:
# Total Spent vs Category
px.box(clean_df, x='Category', y='Total Spent', title='Total Spent by Category')

### What is the Average Order Value per transaction, and how does it vary by category?

In [49]:
average_= clean_df['Total Spent'].describe()['mean']
print(f'Average Total Spent: ${average_:,.2f}')

Average Total Spent: $129.65


In [50]:
avg = clean_df.groupby('Category')['Total Spent'].mean().sort_values(ascending=False).reset_index()
fig = px.bar(avg, x='Category', y='Total Spent', title='Average Spent by Category') 
fig.update_yaxes(title_text='Average Spent')
fig.update_traces(hovertemplate='Category: %{x}<br>Average Total Spent: $%{y:,.2f}')
fig.show()



### Highest spending customers?

In [51]:
cust_spent = clean_df.groupby('Customer ID')['Total Spent'].sum().sort_values(ascending=False).reset_index()
px.bar(cust_spent, x='Customer ID', y='Total Spent', title='Customers by Total Spent')

### **Seasonality Analysis**: Are there specific times of the year, month, or week when sales peak?

In [52]:
# revenue over time (day level)
revenue_time = clean_df.groupby('Transaction Date')['Total Spent'].sum().reset_index()
px.line(revenue_time, x='Transaction Date', y='Total Spent', title='Revenue Over Time')

In [53]:
# revenue over time (month level)
revenue_time = clean_df.groupby('month_year')['Total Spent'].sum().reset_index()
px.line(revenue_time, x='month_year', y='Total Spent', title='Revenue Over Time')

In [54]:
px.histogram(clean_df, x='day_of_week', y='Total Spent', color='day_of_week', title='Total Spent by Day of Week')

In [55]:
cat_time = clean_df.groupby(['month', 'Category'])['Total Spent'].sum().reset_index()
px.line(cat_time, x='month', y='Total Spent', color='Category')

### How does the payment method affect the total revenue?

In [56]:
px.pie(clean_df, names='Payment Method', title='Distribution of Payment Methods', values='Total Spent')

### Which location contributes the most to the total revenue?

In [57]:
px.pie(clean_df, names='Location', title='Distribution of Locations', values='Total Spent')

### Is there a relationship between the quantity sold and the price of the items?

In [58]:
px.scatter(clean_df, x='Quantity', y='Price Per Unit', title='Quantity vs Price Per Unit')

In [59]:
px.strip(clean_df, x='Quantity', y='Total Spent')

# 7 **Deployment (Streamlit)**

In [85]:
%%writefile app.py 

#import libraries
import streamlit as st 
import numpy as np
import pandas as pd 
import plotly.express as px 
import plotly.io as pio
pio.templates.default = "plotly"

#load data
df = pd.read_csv('clean_df.csv')

st.set_page_config(page_title="Retail Dashboard", layout="wide")

st.title("📊 Retail Transactions Dashboard")


st.set_page_config(layout="wide")


# --- Sidebar ---
with st.sidebar:
    
    page = st.radio(
        "📊 Retail Dashboard",
        ["🏠 Overview", "📈 Exploratory Analysis", "💡 Findings & Recommendations"]
    )

    st.markdown("---")
    

# -------------------------------
# ensure datetime format
# -------------------------------
df["Transaction Date"] = pd.to_datetime(df["Transaction Date"])
filtered_df = df.copy()


# -------------------------------
# Sidebar UI
# -------------------------------
st.sidebar.markdown("## 🔍 Filters")

# 1. Date filter    
min_date = df["Transaction Date"].min()
max_date = df["Transaction Date"].max()

date_range = st.sidebar.date_input(
        "📅 Transaction Date",
        value=(min_date, max_date),
        min_value=min_date,
        max_value=max_date
    )
#Apply Date filter
if len(date_range) == 2:
    start, end = pd.to_datetime(date_range)
    filtered_df = filtered_df[
        (filtered_df["Transaction Date"] >= start) &
        (filtered_df["Transaction Date"] <= end)
    ]

# 2. Category filter
category_options = df["Category"].unique()
category = st.sidebar.multiselect("📂 Category", category_options)
# Apply Category filter
if category:
    filtered_df = filtered_df[filtered_df["Category"].isin(category)]

# 3. Item filter (depends on category)

item_options = df[df["Category"].isin(category)]["Item"].unique()
item = st.sidebar.multiselect("🛒 Item", item_options)
# Apply Item filter
if item:
    filtered_df = filtered_df[filtered_df["Item"].isin(item)]

# 4. Payment Method filter
payment_options = df["Payment Method"].unique() 
payment = st.sidebar.multiselect("💳 Payment Method", payment_options)
# Apply Payment filter
if payment:
    filtered_df = filtered_df[filtered_df["Payment Method"].isin(payment)]

# 5. Location filter
location_options = df["Location"].unique()
location = st.sidebar.multiselect("📍 Location", location_options)
# Apply Location filter
if location:
    filtered_df = filtered_df[filtered_df["Location"].isin(location)]


st.sidebar.markdown("---")

st.sidebar.caption("Built by Hussein Ahmed | Epsilon AI Academy")



# --- Routing ---
if page == "🏠 Overview":
    st.title("Overview")
    
    st.subheader("📊 Key Performance Indicators")
    transactions_contribution = (filtered_df.shape[0] / df.shape[0]) * 100
    total_transactions = filtered_df.shape[0]
    total_revenue = filtered_df['Total Spent'].sum()
    avg_order_value = filtered_df['Total Spent'].mean()
    col1, col2, col3, col4 = st.columns(4)

    col1.metric("Transactions Contribution", f"{transactions_contribution:.2f}%")
    col2.metric("Total Transactions", f"{total_transactions:,}")
    col3.metric("Total Revenue", f"${total_revenue:,.2f}")
    col4.metric("Avg Order Value", f"${avg_order_value:,.2f}")

    st.markdown("---")

    top_customer = filtered_df.groupby('Customer ID')['Total Spent'].count().sort_values(ascending=False).reset_index().iloc[0]
    top_customer_name = top_customer['Customer ID']
    top_category = filtered_df.groupby('Category')['Total Spent'].sum().sort_values(ascending=False).reset_index().iloc[0]
    top_category_name = top_category['Category']
    top_item = filtered_df.groupby('Item')['Quantity'].sum().sort_values(ascending=False).reset_index().iloc[0]
    top_item_name = top_item['Item']    
    col5, col6, col7 = st.columns(3)

    col5.metric("Top Customer", top_customer_name)
    col6.metric("Most Profitable Category", top_category_name)
    col7.metric("Best Selling Item", top_item_name)
    st.markdown("---")
    
    # Dataframe view
    st.dataframe(filtered_df)
    st.write(filtered_df.shape)

elif page == "📈 Exploratory Analysis":
    st.title("Exploratory Analysis")
    tab1, tab2, tab3 = st.tabs([
        "🔢 Numerical Analysis",
        "📊 Categorical Analysis",
        "⏱️ Time-Based Analysis"
        ])

    # =========================
    # 🔢 Numerical
    # =========================
    with tab1:
        st.markdown("## 🔢 Numerical Analysis")
        st.markdown("Explore distributions and relationships between numerical variables.")

        st.divider()

        st.markdown("### 📈 Distribution of Total Spent")
        fig11 = px.histogram(filtered_df, x="Total Spent", title="Distribution of Total Spent", marginal="box")
        st.plotly_chart(fig11)

        st.markdown("### 🔗 Relationship: Total Spent vs Quantity")
        fig12 = px.strip(filtered_df, x='Quantity', y='Total Spent')
        st.plotly_chart(fig12)

        
        st.markdown("### 🔗 Correlation Matrix")
        corr_matrix = filtered_df[['Price Per Unit', 'Total Spent', 'Quantity']].corr()
        fig13 = px.imshow(corr_matrix, text_auto=True, aspect="auto",color_continuous_scale='Blues')
        st.plotly_chart(fig13)

    # =========================
    # 📊 Categorical
    # =========================
    with tab2:
        st.markdown("## 📊 Categorical Analysis")
        st.markdown("Compare categories and understand revenue contribution.")

        st.divider()

        st.markdown("### 🏷️ Revenue by Category")
        top_cat = filtered_df.groupby(['Discount Applied', 'Category'])['Total Spent'].sum().sort_values(ascending=False).reset_index()
        fig21 = px.histogram(top_cat, x='Category', y='Total Spent', title='Total Spent by Category', color='Discount Applied')
        st.plotly_chart(fig21)

        col21, col22 = st.columns(2)
        with col21:
            st.markdown("### 📍 Location")
            fig22 = px.pie(filtered_df, names='Location', values='Total Spent')
            st.plotly_chart(fig22)
        with col22:
            st.markdown("### 💳 Payment Method")
            fig23 = px.pie(filtered_df, names='Payment Method', values='Total Spent')
            st.plotly_chart(fig23)

        st.markdown("### 🛒 Top Selling Item by Category")
        top_items = filtered_df.groupby(['Category', 'Item'])['Quantity'].sum().sort_values(ascending=False).reset_index()
        columns = ['Item', 'Quantity']
        top_items_cat = pd.DataFrame(columns=columns)

        for category in filtered_df['Category'].unique():
            cat_items = top_items[top_items['Category'] == category].sort_values(by='Quantity', ascending=False).head(1)
            top_items_cat = pd.concat([top_items_cat, cat_items], ignore_index=True)
        top_items_cat = top_items_cat.sort_values(by='Quantity', ascending=False)
        fig24 = px.bar(top_items_cat, x='Item', y='Quantity', title='Top Selling Item by Category', color='Category')
        st.plotly_chart(fig24)

        st.markdown("### 🛒 Top 10 Selling Items in General")
        top_items_gen = filtered_df.groupby(['Category', 'Item'])['Quantity'].sum().sort_values(ascending=False).reset_index().head(10)
        fig25 = px.bar(top_items_gen, x='Item', y='Quantity', color='Category', title='Top Selling Items in General')
        st.plotly_chart(fig25)
    # =========================
    # ⏱️ Time-Based
    # =========================
    with tab3:
        filtered_df['day_of_week'] = filtered_df['Transaction Date'].dt.day_name()
        filtered_df['month_year'] = filtered_df['Transaction Date'].dt.to_period('M').astype(str)
        st.markdown("## ⏱️ Time-Based Analysis")
        st.markdown("Analyze trends, seasonality, and time patterns.")

        st.divider()

        st.markdown("### 📅 Revenue Over Time")
        revenue_time = filtered_df.groupby('Transaction Date')['Total Spent'].sum().reset_index()
        fig31= px.line(revenue_time, x='Transaction Date', y='Total Spent', title='Revenue Over Time')
        st.plotly_chart(fig31)

        st.markdown("### 📊 Monthly Trends")
        revenue_time = filtered_df.groupby('month_year')['Total Spent'].sum().reset_index()
        fig32 = px.line(revenue_time, x='month_year', y='Total Spent', title='Revenue Over Time')
        st.plotly_chart(fig32)

        st.markdown("### 📊 Day of Week Analysis")
        revenue_day = filtered_df.groupby('day_of_week')['Total Spent'].sum().reset_index()
        fig33 = px.bar(revenue_day, x='day_of_week', y='Total Spent', title='Revenue by Day of Week')
        st.plotly_chart(fig33)

elif page == "💡 Findings & Recommendations":
    st.title("Findings & Recommendations")

    col1, col2 = st.columns(2)

    with col1:
        st.markdown("### ✅ Key Findings")
        st.write("""
        - The **Butchers** category contributes the highest share of total revenue.
        ----------------------------------------
        - Revenue does **not vary** significantly across locations or payment methods.
        ----------------------------------------
        - Edible and drinkable categories have higher demand compared to others.
        <br><br>
        ----------------------------------------
        - The most sold item is Beverage 2.
        <br><br><br>
        ----------------------------------------
        - Customer spending is relatively similar, with Customer 24 being the highest spender.
        ----------------------------------------
        - Discounts do not significantly impact sales.
        ----------------------------------------
        - There is a logical positive relationship between quantity and total spending.
        ----------------------------------------
        - January shows a consistent increase in revenue each year, followed by a decline in the next month.
        <br>
        ----------------------------------------
        - Sales increase on Fridays.
        """,unsafe_allow_html=True)

    with col2:
        st.markdown("### 🎯 Recommendations")
        st.write("""
        - Prioritize **availability** for Butchers products.
        ----------------------------------------
        - Focus on **operational efficiency**, not payment optimization.
        ----------------------------------------
        - Expand product variety in these categories.
        - Bundle food + beverage offers to increase basket size.
        ----------------------------------------
        - Ensure constant availability (avoid stockouts).
        - Use it as a traffic driver (e.g., promotions).
        - Cross-sell with complementary items (e.g., snacks).
        ----------------------------------------
        - Target Customer 24 with personalized offers.
        ----------------------------------------
        - Reduce unnecessary discounts → protect margins.
        ----------------------------------------
        - Highlight bundle deals.
        ----------------------------------------
        - Launch major campaigns in January to maximize revenue.
        - Prepare retention strategies for post-January drop.
        ----------------------------------------
        - Run promotions specifically on Fridays.
        - Increase inventory and staffing for peak demand.
        - Launch “Friday special” campaigns.
        """,unsafe_allow_html=True)



Overwriting app.py


In [63]:
clean_df.columns

Index(['Customer ID', 'Category', 'Price Per Unit', 'Quantity', 'Total Spent',
       'Payment Method', 'Location', 'Transaction Date', 'Discount Applied',
       'Item', 'year', 'month', 'day_of_week', 'month_year'],
      dtype='object')

# END

In [61]:
! streamlit run app.py

^C


### upload file (additional)

In [ ]:
'''
# Upload file
uploaded_file = st.file_uploader("Upload your dataset (CSV)", type=["csv"])

if uploaded_file:
    df = pd.read_csv(uploaded_file)
    


else:
    st.warning("Please upload a CSV file to visualize the dashboard.")
'''